src/modules/atomic-queue/AtomicQueue.sol

```diff
@@ -0,0 +1,298 @@
+ * @author crispymangoes
+ */
+contract AtomicQueue is ReentrancyGuard {
+    using SafeTransferLib for ERC20;
+    using Math for uint256;
+
+    // ========================================= STRUCTS =========================================
+
+    /**
+     * @notice Stores request information needed to fulfill a users atomic request.
+     * @param deadline unix timestamp for when request is no longer valid
+     * @param atomicPrice the price in terms of `want` asset the user wants their `offer` assets "sold" at
+     * @dev atomicPrice MUST be in terms of `want` asset decimals.
+     * @param offerAmount the amount of `offer` asset the user wants converted to `want` asset
+     * @param inSolve bool used during solves to prevent duplicate users, and to prevent redoing multiple checks
+     */
+    struct AtomicRequest {
+        uint64 deadline; // deadline to fulfill request
+        uint88 atomicPrice; // In terms of want asset decimals
+        uint96 offerAmount; // The amount of offer asset the user wants to sell.
+        bool inSolve; // Indicates whether this user is currently having their request fulfilled.
+    }
+
+    /**
+     * @notice Used in `viewSolveMetaData` helper function to return data in a clean struct.
+     * @param user the address of the user
+     * @param flags 8 bits indicating the state of the user only the first 4 bits are used XXXX0000
+     *              Either all flags are false(user is solvable) or only 1 is true(an error occurred).
+     *              From right to left
+     *              - 0: indicates user deadline has passed.
+     *              - 1: indicates user request has zero offer amount.
+     *              - 2: indicates user does not have enough shares in wallet.
+     *              - 3: indicates user has not given AtomicQueue approval.
+     * @param offerToSolve the amount of offer asset to solve
+     * @param assetsForWant the amount of assets users want for their offer assets
+     */
+    struct SolveMetaData {
+        address user;
+        uint8 flags;
+        uint256 assetsToOffer;
+        uint256 assetsForWant;
+    }
+
+    // ========================================= GLOBAL STATE =========================================
+
+    /**
+     * @notice Maps user address to offer asset to want asset to a AtomicRequest struct.
+     */
+    mapping(address => mapping(ERC20 => mapping(ERC20 => AtomicRequest))) public userAtomicRequest;
+
+    //============================== ERRORS ===============================
+
+    error AtomicQueue__UserRepeated(address user);
+    error AtomicQueue__RequestDeadlineExceeded(address user);
+    error AtomicQueue__UserNotInSolve(address user);
+    error AtomicQueue__ZeroOfferAmount(address user);
+
+    //============================== EVENTS ===============================
+
+    /**
+     * @notice Emitted when `updateAtomicRequest` is called.
+     */
+    event AtomicRequestUpdated(
+        address user,
+        address offer,
```

Comment:
I think it makes sense to change `offer` and `want` to be `offerToken` and `wantToken` because the value type is address. However anywhere we make `offer/want` and ERC20, that should give enough context that we are talking about a token.